python/dgl/sparse/elementwise_op_diag.py

```diff
@@ -78,21 +77,20 @@ def diag_sub(D1: DiagMatrix, D2: DiagMatrix) -> DiagMatrix:
             f"{D1.shape} and D2 {D2.shape} must match."
         )
         return diag(D1.val - D2.val, D1.shape)
-    raise RuntimeError(
-        "Elementwise subtraction between "
-        f"{type(D1)} and {type(D2)} is not supported."
-    )
+    # Python falls back to D2.__rsub__(D1) then TypeError when NotImplemented
+    # is returned.
+    return NotImplemented
 
 
 def diag_mul(
-    D1: Union[DiagMatrix, float, int], D2: Union[DiagMatrix, float, int]
+    D1: DiagMatrix, D2: Union[DiagMatrix, float, int]
```

Comment:
Is this valid since we have both `DiagMatrix.__mul__ = diag_mul` and `DiagMatrix.__rmul__ = diag_mul`?

Reply:
It is.
`D1 * D2 == D1.__mul__(D2) == diag_mul(D1, D2)` works normally.
`D1 * v == D1.__mul__(v) == diag_mul(D1, v)` also works.
`v * D1 == v.__mul__(D1)` will fail initially since the scalar multiplication doesn't support `DiagMatrix`.  Python will then try `D1.__rmul__(v) == diag_mul(D1, v)` which should be correct since multiplication is commutative.
`__rsub__` and `__rtruediv__` should have different routines from `__sub__` and `__truediv__` though, which is why I removed them in this PR.